v2/spanner-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/transforms/AssignShardIdFn.java

```diff
@@ -231,7 +231,10 @@ public void processElement(ProcessContext c) throws Exception {
 
       record.setShard(qualifiedShard);
       String finalKeyString = tableName + "_" + keysJsonStr + "_" + qualifiedShard;
-      Long finalKey = finalKeyString.hashCode() % maxConnectionsPerShard;
+      Long finalKey =
+          finalKeyString.hashCode()
+              % maxConnectionsAcrossAllShards; // The total parallelism is
+                                               // maxConnectionsAcrossAllShards
```

Comment:
Question: I remember discussing this guarantees ordering per key since Dataflow processes a thread in the same order. But is it a guarantee by dataflow? Example lets say a thread restarts when it had (key1, write1) followed by (key1, write2), will this order be retained after the thread restarts?

Otherwise in these cases, it could happen the MySQL write went through but the spanner write was not executed. This could lead to inconsistencies.

Reply:
Yeah, as per this: https://g3doc.corp.google.com/company/gfw/support/cloud/playbooks/dataflow/streaming_pipeline_slow_stuck.md?cl=head#keys-in-dataflow - there is a guarantee by Dataflow. For your scenario about worker restart - yes this is possible that for a given element - in mid-processing, the worker restarted. Then Dataflow will give the same item to different worker thread - but it will not happen that two threads are processing the same element in parallel. So the half-processed element will begin it's processing again - since there was no update done to Spanner shadow table - it will do upsert to MySQL and this time the Spanner write will also happen. 

There is a second scenario - let's suppose there is only one Dataflow key in the system, for simplicity. It can happen that MySQL write succeeded and the Spanner shadow table write failed - some Spanner issue - then this record gets written to DLQ for reprocessing. Meanwhile, it can happen that data with an earlier commit timestamp arrives - we will end up writing this to MySQL and update the commit timestamp in Spanner. After DLQ record is processed,MySQL will have new data again. So in essence: the source is eventually consistent (by design)